plex/player.py

```diff
@@ -0,0 +1,218 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+#  Copyright (C) 2019 Sascha Montellese <montellese@kodi.tv>
+#
+#  SPDX-License-Identifier: GPL-2.0-or-later
+#  See LICENSES/README.md for more information.
+#
+import time
+import threading
+
+import xbmc
+import xbmcmediaimport
+
+from plex.api import Api
+from plex.constants import PLEX_PROTOCOL, PLEX_PLAYER_PLAYING, PLEX_PLAYER_PAUSED, PLEX_PLAYER_STOPPED
+from plex.server import Server
+
+from lib.utils import log, mediaProvider2str, toMilliseconds
+
+REPORTING_INTERVAL = 5 # seconds
+
+class Player(xbmc.Player):
+    def __init__(self):
+        '''Initializes the player'''
+        super(Player, self).__init__()
+
+        self._providers = {}
+        self._lock = threading.RLock()
+
+        self._state = {'playbacktime': None, 'state': None, 'lastreport': None}
+
+
+    def AddProvider(self, mediaProvider):
+        '''Adds a media provider to the player'''
+        if not mediaProvider:
+            raise ValueError('invalid mediaProvider')
+
+        with self._lock:
+            self._providers[mediaProvider.getIdentifier()] = mediaProvider
+
+
+    def RemoveProvider(self, mediaProvider):
+        '''Removes the associated media provider'''
+        if not mediaProvider:
+            raise ValueError('invalid mediaProvider')
+
+        with self._lock:
+            del self._providers[mediaProvider.getIdentifier()]
+
+
+    def Process(self):
+        '''Called from the observer thread to periodically report the state to PMS'''
+        with self._lock:
+            if self.isPlaying():
+                lastreport = self._state.get('lastreport')
+
+                if not lastreport:
+                    return
+
+                if (time.time() - lastreport) < REPORTING_INTERVAL:
+                    return
+
+                if self._item:
+                    self._syncPlaybackState(playbackTime=self._getPlayingTime())
+
+
+    def onPlayBackStarted(self):
+        '''Triggered when xbmc.Player is started'''
+        self._reset()
```

Comment:
I would take the lock here to make sure that no other thread can come between executing `self._reset()` and `self._getPlayingFile()`. Otherwise you might end up with this thread executing `self._reset()` then another thread does something and then this thread executes `self._getPlayingFile()` which might again end up with an invalid state.

The same applies to all other methods. Then you can get rid of the locks in the private methods and use a normal `Lock` again.